core/src/main/scala/org/apache/spark/internal/io/SparkHadoopWriter.scala

```diff
@@ -0,0 +1,402 @@
+  // Create writer
+  // --------------------------------------------------------------------------
+
+  def initWriter(taskContext: NewTaskAttemptContext, splitId: Int): Unit = {
+    val numfmt = NumberFormat.getInstance(Locale.US)
+    numfmt.setMinimumIntegerDigits(5)
+    numfmt.setGroupingUsed(false)
+
+    val outputName = "part-" + numfmt.format(splitId)
+    val path = FileOutputFormat.getOutputPath(getConf())
+    val fs: FileSystem = {
+      if (path != null) {
+        path.getFileSystem(getConf())
+      } else {
+        FileSystem.get(getConf())
+      }
+    }
+
+    writer = getConf().getOutputFormat
+      .getRecordWriter(fs, getConf(), outputName, Reporter.NULL)
+      .asInstanceOf[RecordWriter[K, V]]
+
+    require(writer != null, "Unable to obtain RecordWriter")
+  }
+
+  def write(pair: (K, V)): Unit = {
+    require(writer != null, "Must call createWriter before write.")
+    writer.write(pair._1, pair._2)
+  }
+
+  def closeWriter(taskContext: NewTaskAttemptContext): Unit = {
+    if (writer != null) {
+      writer.close(Reporter.NULL)
+    }
+  }
+
+  // --------------------------------------------------------------------------
+  // Create OutputFormat
+  // --------------------------------------------------------------------------
+
+  def initOutputFormat(jobContext: NewJobContext): Unit = {
+    if (outputFormat == null) {
+      outputFormat = getConf().getOutputFormat.getClass
+        .asInstanceOf[Class[_ <: OutputFormat[K, V]]]
+    }
+  }
+
+  private def getOutputFormat(): OutputFormat[K, V] = {
+    require(outputFormat != null, "Must call initOutputFormat first.")
+
+    outputFormat.newInstance()
+  }
+
+  // --------------------------------------------------------------------------
+  // Verify hadoop config
+  // --------------------------------------------------------------------------
+
+  def assertConf(): Unit = {
+    val outputFormatInstance = getOutputFormat()
+    val keyClass = getConf().getOutputKeyClass
+    val valueClass = getConf().getOutputValueClass
+    if (outputFormatInstance == null) {
+      throw new SparkException("Output format class not set")
+    }
+    if (keyClass == null) {
+      throw new SparkException("Output key class not set")
+    }
+    if (valueClass == null) {
+      throw new SparkException("Output value class not set")
+    }
+    SparkHadoopUtil.get.addCredentials(getConf())
+
+    logDebug("Saving as hadoop file of type (" + keyClass.getSimpleName + ", " +
+      valueClass.getSimpleName + ")")
+  }
+
+  def checkOutputSpecs(jobContext: NewJobContext): Unit = {
+    val ignoredFs = FileSystem.get(getConf())
+    getOutputFormat().checkOutputSpecs(ignoredFs, getConf())
+  }
+
+}
+
+/**
+ * A helper class that reads Configuration from newer mapreduce API, creates output
+ * Format/Committer/Writer.
+ */
+private[spark]
+class SparkHadoopMapReduceWriterConfig[K, V: ClassTag](conf: SerializableConfiguration)
+  extends SparkHadoopWriterConfig[K, V] with Logging {
```

Comment:
This supports the new `mapreduce` API, which creates OutputFormat from `jobContext`.